keys.go

```diff
@@ -0,0 +1,174 @@
+// Returns the server keys and the state of the TLS connection used to retrive them.
+func FetchKeysDirect(serverName, addr, sni string) (*ServerKeys, *tls.ConnectionState, error) {
+	// Create a TLS connection.
+	tcpconn, err := net.Dial("tcp", addr)
+	if err != nil {
+		return nil, nil, err
+	}
+	defer tcpconn.Close()
+	tlsconn := tls.Client(tcpconn, &tls.Config{ServerName: sni})
+	err = tlsconn.Handshake()
+	if err != nil {
+		return nil, nil, err
+	}
+	connectionState := tlsconn.ConnectionState()
+
+	// Write a GET /_matrix/key/v2/server down the connection.
+	requestURL := "matrix://" + serverName + "/_matrix/key/v2/server"
+	request, err := http.NewRequest("GET", requestURL, nil)
+	if err != nil {
+		return nil, nil, err
+	}
+	request.Header.Set("Connection", "close")
+	if err = request.Write(tlsconn); err != nil {
+		return nil, nil, err
+	}
+
+	// Read the 200 OK from the server.
+	response, err := http.ReadResponse(bufio.NewReader(tlsconn), request)
+	if err != nil {
+		return nil, nil, err
+	}
+	var keys ServerKeys
+	if keys.Raw, err = ioutil.ReadAll(response.Body); err != nil {
+		return nil, nil, err
+	}
+	if err = json.Unmarshal(keys.Raw, &keys); err != nil {
+		return nil, nil, err
+	}
+	return &keys, &connectionState, nil
+}
+
+// KeyChecks are the checks that should be applied to ServerKey responses.
+type KeyChecks struct {
+	AllChecksOK               bool // Did all the checks pass?
+	MatchingServerName        bool // Does the server name match what was requested.
+	FutureValidUntilTS        bool // The valid until TS is in the future.
+	HasEd25519VerifyKey       bool // The key response has a E25519 key for the server.
+	ValidEd25519VerifyKey     bool // The verify keys are value.
+	MatchingEd25519Signature  bool // Every verify key claimed has a valid signature.
+	HasTLSFingerprint         bool // The response includes a TLS fingerprint.
+	ValidSHA256TLSFingerprint bool // Every TLS fingerprint includes a SHA-256 hash.
+	MatchingTLSFingerprint    bool // The TLS fingerprint for the connection matches one of the listed fingerprints.
+}
+
+// CheckKeys checks the keys returned from a server to make sure they are valid.
+// If the checks pass the also returns a map of key_id to Ed25519 public key and a list of SHA256 TLS fingerprints.
+func CheckKeys(serverName string, timeNowMs uint64, keys ServerKeys, connState *tls.ConnectionState) (
+	checks KeyChecks, ed25519Keys map[string]Base64String, sha256Fingerprints []Base64String,
+) {
+	checks.MatchingServerName = serverName == keys.ServerName
+
+	ed25519Keys = checkVerifyKeys(keys, &checks)
+	sha256Fingerprints = checkTLSFingerprints(keys, &checks)
+
+	checks.FutureValidUntilTS = timeNowMs < keys.ValidUntilTS
+
+	if connState == nil {
+		// No TLS connection state so skip the fingerprint checks.
+		checks.MatchingTLSFingerprint = true
+	} else {
+		// Check the peer certificates.
+		checks.MatchingTLSFingerprint = checkFingerprint(connState, sha256Fingerprints)
```

Comment:
I'd rather not use a nil/empty slice for no certificates here since someone might unpack a zero value ConnectionState and end up passing an empty list by mistake.